# MMB-117 Environmental microbiology - lab course 2019

## Thursday 09.05.19: Microbial community data analysis in R

**Dr. Igor S. Pessi**  
*Arctic Microbial Ecology*  
*Department of Microbiology*  
*[igor.pessi@helsinki.fi](mailto:igor.pessi@helsinki.fi)*

After quality trimming, chimera removal, OTU clustering and taxonomic assignment we have transformed the data into a biom file that contains the OTU table and taxonomic assignment for each OTU.

In this tutorial we will continue our analysis of 16S rRNA gene data using the R package **phyloseq**. This will be a very basic overview, more information and many good tutorials can be found in their website [https://joey711.github.io/phyloseq/index.html](https://joey711.github.io/phyloseq/index.html).

### Installing R and RStudio

First we need to install R. Just download below the appropriate version for your computer and follow the instructions:

* Windows: [https://cran.rstudio.com/bin/windows/base/R-3.5.0-win.exe](https://cran.rstudio.com/bin/windows/base/R-3.5.0-win.exe)
* Mac OS X 10.6 (Snow Leopard) – 10.8 (Mountain Lion):[https://cran.rstudio.com/bin/macosx/R-3.2.1-snowleopard.pkg](https://cran.rstudio.com/bin/macosx/R-3.2.1-snowleopard.pkg)
* Mac OS X 10.9 (Mavericks) and higher: [https://cran.rstudio.com/bin/macosx/R-3.3.3.pkg](https://cran.rstudio.com/bin/macosx/R-3.3.3.pkg)
* Mac OS X 10.11 (El Capitan) and higher [https://cran.rstudio.com/bin/macosx/R-3.5.0.pkg](https://cran.rstudio.com/bin/macosx/R-3.5.0.pkg)

Once R has finished installing, we can then install RStudio:

* Windows Vista/7/8/10: [https://download1.rstudio.org/RStudio-1.1.447.exe](https://download1.rstudio.org/RStudio-1.1.447.exe)
* Mac OS X 10.6+ (64-bit): [https://download1.rstudio.org/RStudio-1.1.447.dmg](https://download1.rstudio.org/RStudio-1.1.447.dmg)

### Installing R packages

From the book [**R packages**](http://r-pkgs.had.co.nz/intro.html):

> In R, the fundamental unit of shareable code is the package. A package bundles together code, data, documentation, and tests, and is easy to share with others. As of January 2015, there were over 6,000 packages available on the Comprehensive R Archive Network, or CRAN, the public clearing house for R packages. This huge variety of packages is one of the reasons that R is so successful: the chances are that someone has already solved a problem that you’re working on, and you can benefit from their work by downloading their package.

So let's install the packages we will use in our analyses. Open RStudio and type in the console:

```
# phyloseq
install.packages("BiocManager")
BiocManager::install("phyloseq")

# ggplot2
install.packages("ggplot2")

# vegan
install.packages("vegan")
```

### Importing data and preprocessing

**DISCLAIMER:** The only way to become proficient in R is by typing, making mistakes and trying again. So I strongly advise that in the examples given below you type the commands instead of copying and pasting them. You have been warned!

**DISCLAIMER 2:** R (as many other programming/scripting languages) is *case-sensitive*. That means that **a** and **A** are treated as different symbols, so pay attention when you’re typing!

<!-- We will start by doing some basic operations to get a feeling of how R works. We start by, of course, launching RStudio.

We will now do some simple calculations and assignments. In the RStudio console, type and press ENTER:

```
3*5+10
```

Take a look at the ouput. What has happened? Pretty obvious, no?

Now we will calculate the area of a circle with a radius of 10. For this, we will first create an object called **radius**, and assign it the value of 10. Then, we will calculate the area of a circle and store the results on another object we will call **area**. In the console, type (one line at a time):

```
radius <- 10
area <- - pi * radius^2
```

Take a look at the *Environment* tab in the top-right of the RStudio window. What has changed after you typed in the commands above? Differently from the first example, here we gave R a command and told it to store the results in an object instead of printing it to the screen. To take a look at these objects we just created, simply type their name in the console:

```
radius
area
```

What has the command **area <- pi * radius^2** done? Can you understand how the syntax works? What does the characters **\*** and **^2** denote? Finally, what would happen if you assign another value to the object **radius**? Try it and see it for yourself!

We will now do some slightly more complicated operations. As it is mentioned in the disclaimer above, learning R is a bit of a steep curve so don’t be upset if you don’t understand 100% of what is happening. This brief introduction is intended to  give you an overview of the capabilities of R and hopefully will entice you to keep learning!

Now for some real 16S analyses! -->

First we need to tell R where our working directory is located. This is the folder where you have put the biom and metadata files. In the RStudio menu bar (on the top of the window), click in *Session > Set Working Directory > Choose Directory*. In the window that popped up, navigate to the folder, select it and click **Open**.

The next step now is to load the newly-installed packages:

```
library("phyloseq")
library("ggplot2")
library("vegan")
```

Now we have phyloseq hopefully working and we can move on reading in the data. Take care that the paths to the files and all filenames are correct. The biom fle can be read in to R with the function ìmport_biom().

```
mydata <- import_biom("otu_table_added_taxa.biom")
```

Let's investigate the object **mydata**. For that, just type:

```
mydata
```

To take a look at the OTU and taxonomy tables, we can do:

```
otu_table(mydata)
tax_table(mydata)
```

The way our taxonomy table is formatted will cause some downstream errors:

```
Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
       Rank1           Rank2                 Rank3                     
OTU68  "Bacteria(100)" "Proteobacteria(100)" "Alphaproteobacteria(100)"
OTU9   "Bacteria(100)" "Proteobacteria(100)" "Gammaproteobacteria(100)"
OTU1   "Bacteria(100)" "Proteobacteria(100)" "Alphaproteobacteria(100)"
OTU633 "Bacteria(100)" "Acidobacteria(100)"  "Solibacteres(100)"       
OTU41  "Bacteria(100)" "Acidobacteria(100)"  "Subgroup_2(100)"    
```

To continue we need to remove those confidence values assigned to each taxonomic level (the numbers inside parentheses). To do this, type:

```
tax_table(mydata) <- gsub("\\([0-9]{2,3}\\)", "", tax_table(mydata))
```

The naming of the taxonomic ranks are a bit non-informative:

```
colnames(tax_table(mydata))
```

Let's make them more informative:

```
colnames(tax_table(mydata)) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
```

Let's now remove the negative control from our dataset:

```
mydata <- prune_samples(sample_names(mydata)[1:10], mydata)
```

And for last but not least, let's load in the metadata:

```
metadata <- read.table("metadata.txt", sep="\t", header=TRUE, row.names=1)
sample_data(mydata) <- sample_data(metadata)
```

Now that we have a phyloseq object with all the data we need, we can have a look at it and understand how the different parts can be extracted from the object. Just by giving the name of the object you see the different objects wrapped inside it. It also tells you the the number of samples, taxa and metadata variables. Also you can see the functions to call the different objects.

```
mydata
```

For example, to take a look at the first rows of the taxonomy table, and then the metadata:

```
head(tax_table(mydata))
sample_data(mydata)
```

We already removed some unwanted sequences with mothur and that can also be done with phyloseq. There might be still some Chloroplast sequences in the data, so we should remove them.

```
# Check if there are any chloroplast sequences
subset_taxa(mydata, Class=="Chloroplast")

# And then remove them
mydata <- subset_taxa(mydata, Class!="Chloroplast")
```

### Library sizes and normalisation

Not all samples have the same amount of sequences or, in another words, we have different library sizes and that might influence our results. There are several ways to normalise the data and we will use the most simple one, proportions. Calculate the library sizes of your samples with the function **sample_sums()**, and use **barplot()** to visualize the library size differences.

```
lib_sizes <- sample_sums(mydata)
barplot(lib_sizes)
```

Even though the differences are rather small, it is still a good idea to transform the OTU counts into proportions:

```
mydata_prop <- transform_sample_counts(mydata, function(x) x/sum(x))
lib_sizes <- sample_sums(mydata_prop)
barplot(lib_sizes)
```

### Few figures from the data

We can also use phyloseq to calculate and visualise some α- and β- diversity measures. Calculate Shannon and inverse Simpson indeces using the raw count data:

```
plot_richness(mydata)
```

Then β-diversity with Bray-Curtis dissimilarity index and visualise it with PCoA ordination. In here use the normalised data. You can also do another one with raw counts and see if there is any difference. Use the pH values from the metadata to color the points.

```
mydata_ord <- ordinate(mydata_prop, method="PCoA", distance="bray")
plot_ordination(mydata_prop, mydata_ord, col="ph", title="Normalised counts")
```

And again with the non-normalised data:

```
mydata_ord <- ordinate(mydata, method="PCoA", distance="bray")
plot_ordination(mydata, mydata_ord, col="ph", title="Raw counts")
```

```
mydata_order <- tax_glom(mydata_prop, taxrank="Order")
mydata_abund <- prune_taxa(names(sort(taxa_sums(mydata_order),TRUE)[1:10]), mydata_order)
plot_bar(mydata_abund, fill="Order")
```

Take a look at the graph. What is/are the most abundant microbial groups in the samples? Are there any differences in microbial community composition between the samples? What are them? Remember to consider the metadata! What are the samples you are analysing from? Do you think that the differences you are seeing are due to what?

### Final remarks

This was a **very** basic introduction to R and how to use it to analyse metagenome data. If you want to learn more, I recommend the material below as good starting points. But remember, you only learn R by using it!

* [https://cran.r-project.org/doc/manuals/r-release/R-intro.pdf](https://cran.r-project.org/doc/manuals/r-release/R-intro.pdf)
* [https://kannu.csc.fi/s/Xk2jIQT7wiB3H1c/download](https://kannu.csc.fi/s/Xk2jIQT7wiB3H1c/download)
* [https://www.rstudio.com/wp-content/uploads/2016/10/r-cheat-sheet-3.pdf](https://www.rstudio.com/wp-content/uploads/2016/10/r-cheat-sheet-3.pdf)
